Doxygen/Docs/man/man3/Object.3
.TH "Object" 3 "Fri Mar 29 2013" "Version 31337" "HyperGrafx" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Object \- 
.PP
\fBObject\fP Class: Renderable \fBObject\fP Implementation\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Object\&.hpp>\fP
.PP
Inherits \fBScene\fP\&.
.PP
Inherited by \fBCamera\fP, and \fBParticleSystem\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBUniforms\fP { \fBBEGIN\fP, \fBIS_TEXTURED\fP =  BEGIN, \fBOBJECT_CTM\fP, \fBMORPH_PCT\fP, \fBTEX_SAMPLER\fP, \fBEND\fP }"
.br
.RI "\fIenum Uniforms describes the properties of the base object that need to be visible to the GPU\&. \fP"
.ti -1c
.RI "typedef const unsigned int \fBUniformEnum\fP"
.br
.RI "\fIThe \fBObject\fP class takes advantage of child-extendible enumerations\&. \fP"
.ti -1c
.RI "typedef std::map
.br
< \fBObject::UniformEnum\fP, 
.br
std::string > \fBUniformMap\fP"
.br
.RI "\fIWe store mappings of Uniform Enumerations, The desired function of the var, to strings, the names of the variables\&. \fP"
.ti -1c
.RI "typedef enum \fBObject::Uniforms\fP \fBUniform\fP"
.br
.RI "\fIenum Uniforms describes the properties of the base object that need to be visible to the GPU\&. \fP"
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBObject\fP (const std::string &\fBname\fP, GLuint gShader)"
.br
.RI "\fIConstructor\&. \fP"
.ti -1c
.RI "virtual \fB~Object\fP (void)"
.br
.RI "\fIDefault destructor\&. \fP"
.ti -1c
.RI "void \fBdraw\fP (void)"
.br
.RI "\fIdraw method: Render this object to the screen _buffer\&. \fP"
.ti -1c
.RI "void \fBbuffer\fP (void)"
.br
.RI "\fIbuffer all of our data: Vertices, TexUVs, Normals, Indices, Colors and Morph Buffers\&. \fP"
.ti -1c
.RI "void \fBbufferMorphOnly\fP (void)"
.br
.RI "\fIbuffer only the Morph-related buffers\&. \fP"
.ti -1c
.RI "void \fBdrawMode\fP (GLenum new_mode)"
.br
.RI "\fISelect a new OpenGL draw mode for this \fBObject\fP\&. \fP"
.ti -1c
.RI "void \fBtexture\fP (const char **filename)"
.br
.RI "\fIFIXME: This is a junk, nonflexible method\&. \fP"
.ti -1c
.RI "const std::string & \fBname\fP (void) const "
.br
.RI "\fIRetrieve the _name of this \fBObject\fP\&. \fP"
.ti -1c
.RI "virtual void \fBlink\fP (\fBUniformEnum\fP which, const std::string &\fBname\fP)"
.br
.RI "\fIlink a specified Uniform against the shader's variable _name\&. \fP"
.ti -1c
.RI "virtual void \fBsend\fP (\fBUniformEnum\fP which)"
.br
.RI "\fISend a Uniform to the shader\&. \fP"
.ti -1c
.RI "virtual GLuint \fBshader\fP (void)"
.br
.RI "\fIReturns the \fBObject\fP's current shader\&. \fP"
.ti -1c
.RI "virtual void \fBshader\fP (GLuint newShader)"
.br
.RI "\fISets the shader to be used by this object\&. \fP"
.ti -1c
.RI "void \fBanimation\fP (void(*anim_func)(\fBTransCache\fP &arg))"
.br
.RI "\fIApply an animation callback function to this \fBObject\fP\&. \fP"
.ti -1c
.RI "void \fBpropagate\fP (void)"
.br
.RI "\fIScene-graph changes are not automatically applied to children\&. \fP"
.ti -1c
.RI "\fBvec4\fP \fBposition\fP () const "
.br
.RI "\fIObtain the vec4 representative of the \fBObject\fP's current position in space\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBmorphTarget\fP () const "
.br
.RI "\fIRetrieve a pointer to this object's morph target\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBgenMorphTarget\fP (GLuint \fBshader\fP)"
.br
.RI "\fIInstantiate a new morphing target\&. \fP"
.ti -1c
.RI "float \fBmorphPercentage\fP () const "
.br
.RI "\fIRetrieve the morph Percentage of this object\&. \fP"
.ti -1c
.RI "void \fBmorphPercentage\fP (const float newPercentage)"
.br
.RI "\fISet the morph percentage of this \fBObject\fP\&. \fP"
.ti -1c
.RI "void \fBdestroyMorphTarget\fP ()"
.br
.RI "\fIObliterate the morph target for this object\&. \fP"
.ti -1c
.RI "int \fBnumberOfPoints\fP ()"
.br
.RI "\fIRetrieve the number of _vertices this object has\&. \fP"
.ti -1c
.RI "void \fBtextureID\fP (GLint newTextureID)"
.br
.RI "\fISet the \fBTexture\fP ID / \fBTexture\fP Unit for this \fBObject\fP\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBaddObject\fP (const std::string &objName, GLuint Object_Shader=0)"
.br
.RI "\fIaddObject creates a new \fBObject\fP with the given name and, optionally, a specified shader and adds it to the \fBScene\fP graph\&. \fP"
.ti -1c
.RI "void \fBdelObject\fP (const std::string &objName)"
.br
.RI "\fIdelObject will remove from the \fBScene\fP graph the object with the given name\&. \fP"
.ti -1c
.RI "void \fBdelObject\fP (void)"
.br
.RI "\fIdelObject with no parameters will delete the first \fBObject\fP in the \fBScene\fP\&. \fP"
.ti -1c
.RI "void \fBpopObject\fP (void)"
.br
.RI "\fIpopObject deletes the last \fBObject\fP in the \fBScene\fP\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fBnext\fP (void)"
.br
.ti -1c
.RI "\fBObject\fP * \fBprev\fP (void)"
.br
.ti -1c
.RI "\fBObject\fP * \fBactive\fP (void) const "
.br
.ti -1c
.RI "\fBObject\fP * \fBoperator[]\fP (const std::string &objname)"
.br
.ti -1c
.RI "void \fBinsertObject\fP (\fBObject\fP *obj)"
.br
.RI "\fIRegister a created object with the scene graph\&. \fP"
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "std::vector< \fBAngel::vec4\fP > \fB_vertices\fP"
.br
.RI "\fIvertex buffer\&. \fP"
.ti -1c
.RI "std::vector< \fBAngel::vec3\fP > \fB_normals\fP"
.br
.RI "\fINormals buffer\&. \fP"
.ti -1c
.RI "std::vector< unsigned int > \fB_indices\fP"
.br
.RI "\fIDraw Order Index buffer\&. \fP"
.ti -1c
.RI "std::vector< \fBAngel::vec4\fP > \fB_colors\fP"
.br
.RI "\fIColors buffer\&. \fP"
.ti -1c
.RI "std::vector< \fBAngel::vec2\fP > \fB_texUVs\fP"
.br
.RI "\fI\fBTexture\fP Coordinates buffer\&. \fP"
.ti -1c
.RI "\fBTransCache\fP \fB_trans\fP"
.br
.RI "\fIThe _trans cache encompasses the current transformational state of this object\&. \fP"
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "void \fBdeleteObject\fP (\fBObject\fP *obj)"
.br
.RI "\fIVery seriously delete a child object and free his memory\&. \fP"
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "std::string \fB_name\fP"
.br
.RI "\fI_name is used as an identifying handle for the object\&. \fP"
.ti -1c
.RI "GLuint \fB_vao\fP"
.br
.RI "\fIVertex Array \fBObject\fP handle identifying our buffers/object\&. \fP"
.ti -1c
.RI "GLuint \fB_buffer\fP [\fBNUM_BUFFERS\fP]"
.br
.RI "\fIHandles to our buffers (Vertices, TexUVs, etc\&.) \fP"
.ti -1c
.RI "GLenum \fB_drawMode\fP"
.br
.RI "\fIDrawing mode for this object\&. \fP"
.ti -1c
.RI "bool \fB_isTextured\fP"
.br
.RI "\fIIs this object textured? \fP"
.ti -1c
.RI "float \fB_morphPercentage\fP"
.br
.RI "\fIThe percentage of the morph\&. \fP"
.ti -1c
.RI "\fBObject\fP * \fB_morphTarget\fP"
.br
.RI "\fIA pointer to the object we wish to morph into\&. \fP"
.ti -1c
.RI "std::map< \fBObject::UniformEnum\fP, 
.br
std::string > \fB_uniformMap\fP"
.br
.RI "\fIA map between Uniform variable functions and the actual uniform variable names\&. \fP"
.ti -1c
.RI "std::vector< GLint > \fB_handles\fP"
.br
.RI "\fIHandles to Uniforms on the shader\&. \fP"
.ti -1c
.RI "GLint \fB_textureID\fP"
.br
.RI "\fIThe texture unit index this \fBObject\fP uses\&. \fP"
.ti -1c
.RI "std::list< \fBObject\fP * > \fB_list\fP"
.br
.RI "\fIFor the purposes of rapid propagation of scene-graph changes, \fBObject\fP pointers are stored in a regular flat list\&. \fP"
.ti -1c
.RI "std::map< std::string, \fBObject\fP * > \fB_map\fP"
.br
.RI "\fIFor the purposes of accessing named objects quickly, though, objects are also re-stored in an associative map\&. \fP"
.ti -1c
.RI "std::list< \fBObject\fP * >::iterator \fB_currentObj\fP"
.br
.RI "\fIWe keep an iterator on-hand that references what the scene considers to be it's active, current object\&. \fP"
.ti -1c
.RI "GLuint \fB_gShader\fP"
.br
.RI "\fIA handle to a shader program to be used as the default shader for new children objects added to the scene\&. \fP"
.in -1c
.SS "Private Types"

.in +1c
.ti -1c
.RI "enum \fBBufferType\fP { \fBVERTICES\fP, \fBNORMALS\fP, \fBINDICES\fP, \fBCOLORS\fP, \fBTEXCOORDS\fP, \fBVERTICES_MORPH\fP, \fBNORMALS_MORPH\fP, \fBCOLORS_MORPH\fP, \fBNUM_BUFFERS\fP }"
.br
.RI "\fIThese enumerations describe the types of buffers we want on the GPU per-object\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
\fBObject\fP Class: Renderable \fBObject\fP Implementation\&. 

\fBAuthor:\fP
.RS 4
John Huston 
.RE
.PP
\fBDate:\fP
.RS 4
2013-03-15
.RE
.PP
The \fBObject\fP class represents 'one renderable object' in terms of the scene graph\&. It is a simple unit that is rendered with a single \fBdraw()\fP call\&.
.PP
It contains all of the necessary state for sending uniforms to the shader, all of the buffers needed to send to the card, and also contains a fully-featured scene graph within itself such that children objects can be attached\&. 
.PP
Definition at line 36 of file Object\&.hpp\&.
.SH "Member Typedef Documentation"
.PP 
.SS "typedef enum \fBObject::Uniforms\fP  \fBObject::Uniform\fP"

.PP
enum Uniforms describes the properties of the base object that need to be visible to the GPU\&. BEGIN and END are special sentinel enumerations that must be first and last, respectively\&. 
.SS "typedef const unsigned int \fBObject::UniformEnum\fP"

.PP
The \fBObject\fP class takes advantage of child-extendible enumerations\&. We create an alias here for sake of ease\&. 
.PP
Definition at line 62 of file Object\&.hpp\&.
.SS "typedef std::map< \fBObject::UniformEnum\fP, std::string > \fBObject::UniformMap\fP"

.PP
We store mappings of Uniform Enumerations, The desired function of the var, to strings, the names of the variables\&. This is utilized if we ever switch this object's shader, so we can re-associate with the correct uniform locations\&. 
.PP
Definition at line 70 of file Object\&.hpp\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBObject::BufferType\fP\fC [private]\fP"

.PP
These enumerations describe the types of buffers we want on the GPU per-object\&. NUM_BUFFERS is a special sentinel enumeration that must always be last\&. 
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIVERTICES \fP\fP
VERTICES\&. 
.TP
\fB\fINORMALS \fP\fP
NORMALS\&. 
.TP
\fB\fIINDICES \fP\fP
INDICES\&. 
.TP
\fB\fICOLORS \fP\fP
COLORS\&. 
.TP
\fB\fITEXCOORDS \fP\fP
TEXCOORDS\&. 
.TP
\fB\fIVERTICES_MORPH \fP\fP
VERTICES_MORPH\&. 
.TP
\fB\fINORMALS_MORPH \fP\fP
NORMALS_MORPH\&. 
.TP
\fB\fICOLORS_MORPH \fP\fP
COLORS_MORPH\&. 
.TP
\fB\fINUM_BUFFERS \fP\fP
NUM_BUFFERS This is a sentinel enumeration\&. 
.PP
Definition at line 44 of file Object\&.hpp\&.
.SS "enum \fBObject::Uniforms\fP"

.PP
enum Uniforms describes the properties of the base object that need to be visible to the GPU\&. BEGIN and END are special sentinel enumerations that must be first and last, respectively\&. 
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIBEGIN \fP\fP
BEGIN\&. 
.TP
\fB\fIIS_TEXTURED \fP\fP
IS_TEXTURED\&. 
.TP
\fB\fIOBJECT_CTM \fP\fP
OBJECT_CTM\&. 
.TP
\fB\fIMORPH_PCT \fP\fP
MORPH_PCT\&. 
.TP
\fB\fITEX_SAMPLER \fP\fP
TEX_SAMPLER\&. 
.TP
\fB\fIEND \fP\fP
END\&. 
.PP
Definition at line 79 of file Object\&.hpp\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Object::Object (const std::string &name, GLuintgShader)"

.PP
Constructor\&. Requires at minimum a _name and a shader handle\&. 
.PP
\fBParameters:\fP
.RS 4
\fIname\fP The _name of this object\&. 
.br
\fIgShader\fP The shader to use to render this object\&. 
.RE
.PP

.PP
Definition at line 29 of file Object\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "\fBObject\fP * Scene::addObject (const std::string &objName, GLuintshader = \fC0\fP)\fC [inherited]\fP"

.PP
addObject creates a new \fBObject\fP with the given name and, optionally, a specified shader and adds it to the \fBScene\fP graph\&. If no shader is given, a default shader MUST have been specified for the \fBScene\fP prior to the call\&.
.PP
\fBParameters:\fP
.RS 4
\fIobjName\fP The name of the new \fBObject\fP to add\&. 
.br
\fIObject_Shader\fP The shader that should be used to render this object\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
A pointer to the new \fBObject\fP\&. 
.RE
.PP

.PP
Definition at line 77 of file Scene\&.cpp\&.
.SS "void Object::animation (void(*)(\fBTransCache\fP &arg)anim_func)"

.PP
Apply an animation callback function to this \fBObject\fP\&. Works once only: Does not save the function or automatically run on idle\&. 
.PP
\fBParameters:\fP
.RS 4
\fIanim_func\fP The transformation/animation function to apply\&. 
.RE
.PP

.PP
Definition at line 443 of file Object\&.cpp\&.
.SS "void Scene::deleteObject (\fBObject\fP *obj)\fC [protected]\fP, \fC [inherited]\fP"

.PP

GraphicsProject/Object.cpp
/**
 * @file Object.cpp
 * @author John Huston
 * @authors Nick St.Pierre, Nick VerVoort, Chris Compton
 * @date 2013-03-15
 * @brief Implementation for the Object class.
 */
#include <string>
#include <vector>
#include "Texture.hpp"
#include <stdexcept>

#include "platform.h"
#include "Engine.hpp"
#include "vec.hpp"
#include "mat.hpp"
#include "Object.hpp"
#include "globals.h"
#include "Timer.hpp"

using Angel::vec4;
using Angel::mat4;

/**
 * Constructor. Requires at minimum a _name and a shader handle.
 * @param name The _name of this object.
 * @param gShader The shader to use to render this object.
 */
Object::Object( const std::string &name, GLuint gShader ) {

  /* The constructor is going to initialize the VAO and a series of VBOs.
   The VAO is our general handle to this collection of VBOs.

   Each VBO contains some component data for how to render the vertex:
   Position, Color, Direction (Normal), texture and draw Order. */

  // Create room for our GLUniform _handles
  if ( DEBUG )
    fprintf( stderr, "\nCreating %d handles for uniforms\n", Object::END );
  _handles.resize( Object::END );

  // Associate this Object with the shader.
  shader( gShader );

  // Set our _name from the constructor...
  this->_name = name;

  /* Initialize our draw mode to GL_TRIANGLES until informed otherwise. */
  _drawMode = GL_TRIANGLES;

  // Get Uniform _handles for the following shader variables
  link( Object::IS_TEXTURED, "fIsTextured" );
  link( Object::OBJECT_CTM, "OTM" );
  link( Object::MORPH_PCT, "morphPercentage" );
  link( Object::TEX_SAMPLER, "sampler" );

  //Default to "Not Textured"
  this->_isTextured = false;

  // Linear Interpolation Demo: Morph Percentage
  this->_morphPercentage = 1.0;
  // Pointer to an Object to Morph to.
  this->_morphTarget = NULL;

  /* Create our VAO, which is our handle to all
   the rest of the following information. */glGenVertexArrays( 1, &_vao );
  glBindVertexArray( _vao );
  GLuint glsl_uniform;

  /* Create (Eight) VBOs: One each for Positions, Colors, Normals,
   Textures, draw Order; Three for Morphing (Position,Colors,Normals.) */glGenBuffers(
      NUM_BUFFERS, _buffer );

  /* Create the Vertex _buffer and link it with the shader. */glBindBuffer(
      GL_ARRAY_BUFFER, _buffer[VERTICES] );
  glsl_uniform = glGetAttribLocation( gShader, "vPosition" );
  glEnableVertexAttribArray( glsl_uniform );
  glVertexAttribPointer( glsl_uniform, 4, GL_FLOAT, GL_FALSE, 0, 0 );

  /* Create the MORPH Vertex _buffer and link it with the shader. */glBindBuffer(
      GL_ARRAY_BUFFER, _buffer[VERTICES_MORPH] );
  glsl_uniform = glGetAttribLocation( gShader, "vPositionMorph" );
  glEnableVertexAttribArray( glsl_uniform );
  glVertexAttribPointer( glsl_uniform, 4, GL_FLOAT, GL_FALSE, 0, 0 );

  /* Create the Normal _buffer and link it with the shader. */glBindBuffer(
      GL_ARRAY_BUFFER, _buffer[NORMALS] );
  glsl_uniform = glGetAttribLocation( gShader, "vNormal" );
  glEnableVertexAttribArray( glsl_uniform );
  glVertexAttribPointer( glsl_uniform, 3, GL_FLOAT, GL_FALSE, 0, 0 );

  /* Create the Normal MORPH _buffer and link it with the shader. */glBindBuffer(
      GL_ARRAY_BUFFER, _buffer[NORMALS_MORPH] );
  glsl_uniform = glGetAttribLocation( gShader, "vNormalMorph" );
  glEnableVertexAttribArray( glsl_uniform );
  glVertexAttribPointer( glsl_uniform, 3, GL_FLOAT, GL_FALSE, 0, 0 );

  /* Create the Color _buffer and link it with the shader. */glBindBuffer(
      GL_ARRAY_BUFFER, _buffer[COLORS] );
  glEnable( GL_BLEND );
  glBlendFunc( GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA );
  glsl_uniform = glGetAttribLocation( gShader, "vColor" );
  glEnableVertexAttribArray( glsl_uniform );
  glVertexAttribPointer( glsl_uniform, 4, GL_FLOAT, GL_FALSE, 0, 0 );

  /* Create the Color Morph _buffer and link it with the shader. */glBindBuffer(
      GL_ARRAY_BUFFER, _buffer[COLORS_MORPH] );
  glBlendFunc( GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA );
  glsl_uniform = glGetAttribLocation( gShader, "vColorMorph" );
  glEnableVertexAttribArray( glsl_uniform );
  glVertexAttribPointer( glsl_uniform, 4, GL_FLOAT, GL_FALSE, 0, 0 );

  /* Create the texture Coordinate _buffer and link it with the shader. */glBindBuffer(
      GL_ARRAY_BUFFER, _buffer[TEXCOORDS] );
  glsl_uniform = glGetAttribLocation( gShader, "vTex" );
  glEnableVertexAttribArray( glsl_uniform );
  glVertexAttribPointer( glsl_uniform, 2, GL_FLOAT, GL_FALSE, 0, 0 );

  if ( DEBUG )
    fprintf( stderr, "buffhandles: %u %u %u %u %u %u %u %u\n",
             _buffer[VERTICES], _buffer[NORMALS], _buffer[COLORS],
             _buffer[TEXCOORDS], _buffer[INDICES], _buffer[VERTICES_MORPH],
             _buffer[NORMALS_MORPH], _buffer[COLORS_MORPH] );

  /* Create the Drawing Order _buffer, but we don't need to link it
   with any uniform,
   because we won't be accessing this data directly. (I.e, the numbers here
   are not important once we are in the Vertex shader. */glBindBuffer(
      GL_ELEMENT_ARRAY_BUFFER, _buffer[INDICES] );

  /* Unset the VAO context. */
  glBindVertexArray( 0 );
}

/**
 * Default destructor.
 */
Object::~Object( void ) {

  destroyMorphTarget();

}

/**
 * draw method: Render this object to the screen _buffer.
 */
void Object::draw( void ) {

  glBindVertexArray( _vao );

  // Check to see if the correct shader program is engaged.
  GLint currShader;
  glGetIntegerv( GL_CURRENT_PROGRAM, &currShader );
  if ( (GLuint) currShader != shader() ) {

    Camera *activeCamera = Engine::instance()->cams()->active();

    //if (DEBUG) std::cerr << "Switching shading context.\n";

    // Set OpenGL to use this object's shader.
    glUseProgram( shader() );

    // Set the Active Camera's shader to the Object's shader.
    activeCamera->shader( shader() );

    // send the Camera's info to the new shader.
    activeCamera->view();
  }

  send( Object::IS_TEXTURED );
  send( Object::OBJECT_CTM );
  send( Object::MORPH_PCT );
  send( Object::TEX_SAMPLER );

  //  this->morphPercentage() == -1.0 ? ; : send( Object::MORPH_PCT );

  /* Are we using a draw order? */
  if ( _indices.size() > 1 ) glDrawElements( _drawMode, _indices.size(),
                                             GL_UNSIGNED_INT, 0 );
  else glDrawArrays( _drawMode, 0, _vertices.size() );

  glBindVertexArray( 0 );

  // draw all of our Children.
  // (With clothes on, pervert.)
  Scene::draw();

}

/**
 * buffer all of our data: Vertices, TexUVs, Normals,
 * Indices, Colors and Morph Buffers.
 */
void Object::buffer( void ) {

  glBindVertexArray( _vao );

  glBindBuffer( GL_ARRAY_BUFFER, _buffer[VERTICES] );
  glBufferData( GL_ARRAY_BUFFER, sizeof(Angel::vec4) * _vertices.size(),
                &(_vertices[0]), GL_STATIC_DRAW );

  glBindBuffer( GL_ARRAY_BUFFER, _buffer[NORMALS] );
  glBufferData( GL_ARRAY_BUFFER, sizeof(Angel::vec3) * _normals.size(),
                &(_normals[0]), GL_STATIC_DRAW );

  glBindBuffer( GL_ARRAY_BUFFER, _buffer[COLORS] );
  glBufferData( GL_ARRAY_BUFFER, sizeof(Angel::vec4) * _colors.size(),
                &(_colors[0]), GL_STATIC_DRAW );

  /* Without the following workaround code,
   Mac OSX will segfault attempting to access
   the texcoordinate buffers on nontextured objects. */
  if ( _texUVs.size() == 0 && _isTextured == false ) {
    _texUVs.push_back( Angel::vec2( -1, -1 ) );
  } else if ( _texUVs.size() > 1 ) {
    /* Yes, this workaround prevents us from having
     textured objects with only one point.
     Oops. */
    _isTextured = true;
  }

  glBindBuffer( GL_ARRAY_BUFFER, _buffer[TEXCOORDS] );
  glBufferData( GL_ARRAY_BUFFER, sizeof(Angel::vec2) * _texUVs.size(),
                (_texUVs.size() ? &(_texUVs[0]) : NULL), GL_STATIC_DRAW );

  glBindBuffer( GL_ELEMENT_ARRAY_BUFFER, _buffer[INDICES] );
  glBufferData( GL_ELEMENT_ARRAY_BUFFER, sizeof(unsigned int) * _indices.size(),
                &(_indices[0]), GL_STATIC_DRAW );

  glBindVertexArray( 0 );

}

/**
 * buffer only the Morph-related buffers.
 */
void Object::bufferMorphOnly( void ) {

  glBindVertexArray( _vao );

  glBindBuffer( GL_ARRAY_BUFFER, _buffer[VERTICES_MORPH] );
  glBufferData( GL_ARRAY_BUFFER,
                sizeof(Angel::vec4) * morphTarget()->_vertices.size(),
                &(morphTarget()->_vertices[0]), GL_STATIC_DRAW );

  glBindBuffer( GL_ARRAY_BUFFER, _buffer[NORMALS_MORPH] );
  glBufferData( GL_ARRAY_BUFFER,
                sizeof(Angel::vec3) * morphTarget()->_normals.size(),
                &(morphTarget()->_normals[0]), GL_STATIC_DRAW );

  glBindBuffer( GL_ARRAY_BUFFER, _buffer[COLORS_MORPH] );
  glBufferData( GL_ARRAY_BUFFER,
                sizeof(Angel::vec4) * morphTarget()->_colors.size(),
                &(morphTarget()->_colors[0]), GL_STATIC_DRAW );

  // #MORPH
  // TODO: MORPH TEXTURES AND INDICIES

  glBindVertexArray( 0 );

}

/**
 * Select a new OpenGL draw mode for this Object.
 * Can be GL_LINES, GL_LINE_LOOP, GL_TRIANGLES, etc.
 * @see http://www.opengl.org/wiki/Primitive
 * @param new_mode The primitive rendering mode to use.
 */
void Object::drawMode( GLenum new_mode ) {
  _drawMode = new_mode;
}

/**
 * FIXME: This is a junk, nonflexible method.
 * It would be better if you didn't think of this as being here.
 *
 * @param filename an array of strings to load textures from.
 */
void Object::texture( const char** filename ) {

  Tick.Tock();
  glBindVertexArray( _vao );

  Texture **textures = (Texture **) malloc( sizeof(Texture*) * 5 );
  textures[0] = new Texture( GL_TEXTURE_2D );
  textures[1] = new Texture( GL_TEXTURE_2D );
  textures[2] = new Texture( GL_TEXTURE_2D );
  textures[3] = new Texture( GL_TEXTURE_2D );
  textures[4] = new Texture( GL_TEXTURE_2D );

  textures[0]->load( filename[0] );
  textures[1]->load( filename[1] );
  textures[2]->load( filename[2] );
  textures[3]->load( filename[3] );
  textures[4]->load( filename[4] );

  textures[0]->buffer();
  textures[1]->buffer();
  textures[2]->buffer();
  textures[3]->buffer();
  textures[4]->buffer();

  textures[0]->bind( GL_TEXTURE0 );
  textures[1]->bind( GL_TEXTURE1 );
  textures[2]->bind( GL_TEXTURE2 );
  textures[3]->bind( GL_TEXTURE3 );
  textures[4]->bind( GL_TEXTURE4 );

  GLuint gSampler0 = glGetUniformLocation( shader(), "gSampler0" );
  glUniform1i( gSampler0, 0 );
  GLuint gSampler1 = glGetUniformLocation( shader(), "gSampler1" );
  glUniform1i( gSampler1, 1 );
  GLuint gSampler2 = glGetUniformLocation( shader(), "gSampler2" );
  glUniform1i( gSampler2, 2 );
  GLuint gSampler3 = glGetUniformLocation( shader(), "gSampler3" );
  glUniform1i( gSampler3, 3 );
  GLuint gSampler4 = glGetUniformLocation( shader(), "gSampler4" );
  glUniform1i( gSampler4, 4 );

  glBindVertexArray( 0 );

  Tick.Tock();
  fprintf( stderr, "took %lu usec to finalize textures.\n", Tick.Delta() );
}

/**
 * Retrieve the _name of this Object.
 * @return The _name of this Object.
 */
const std::string &Object::name( void ) const {

  return _name;

}

/**
 * link a specified Uniform against the shader's variable _name.
 *
 * @param which The Uniform to link.
 * @param name The variable _name on the shader.
 */
void Object::link( UniformEnum which, const std::string &name ) {

  if ( which >= _handles.size() ) {
    fprintf(
        stderr,
        "WARNING: Ignoring request to link a uniform (#%u) beyond our handles array [%lu].\n",
        which, _handles.size() );
    return;
  }

  // Save the link between the Uniform and the Variable _name.
  _uniformMap[which] = name;

  if ( DEBUG )
    fprintf( stderr, "Linking enum[%u] with %s for object %s\n", which,
             name.c_str(), this->_name.c_str() );

  _handles[which] = glGetUniformLocation( shader(), name.c_str() );
  if ( DEBUG )
    fprintf( stderr, "Linking handles[%d] to %s; got %d.\n", which,
             name.c_str(), _handles[which] );

}

/**
 * Send a Uniform to the shader.
 * @param which The uniform to send.
 */
void Object::send( Object::UniformEnum which ) {
  switch ( which ) {

  case Object::IS_TEXTURED:
    glUniform1i( _handles[Object::IS_TEXTURED], (_isTextured) ? 1 : 0 );
    break;

  case Object::OBJECT_CTM:
    glUniformMatrix4fv( _handles[Object::OBJECT_CTM], 1, GL_TRUE,
                        this->_trans.OTM() );
    break;

  case Object::MORPH_PCT:
    glUniform1f( _handles[Object::MORPH_PCT], this->morphPercentage() );

    break;

  case Object::TEX_SAMPLER:
    glUniform1i( _handles[Object::TEX_SAMPLER], _textureID );
    break;

  default:
    throw std::invalid_argument( "Unknown Uniform Handle Enumeration." );
  }
}

/**
 Returns the Object's current shader.
 Defined because C++ will not let you overload an overrided function,
 without re-overloading it in the derived class.

 @return a GLuint handle to the shader program used by this Object.
 **/
GLuint Object::shader( void ) {

  return Scene::shader();

}

/**
 Sets the shader to be used by this object.
 Triggers a query of the shader program,
 for the locations of the Uniform locations
 that the object needs.

 @param newShader a GLuint handle to the shader program to use.

 @return None.
 **/
void Object::shader( GLuint newShader ) {

  // Cache the shader handle for later.
  Scene::shader( newShader );

  // We have to use the program to query the glUniform locations.
  glUseProgram( newShader );

  // Re-link our Uniforms to this shader.
  UniformMap::iterator it;
  for ( it = _uniformMap.begin(); it != _uniformMap.end(); ++it ) {
    link( it->first, it->second );
  }

}

/**
 * Apply an animation callback function to this Object.
 * Works once only: Does not save the function or automatically run on idle.
 * @param anim_func The transformation/animation function to apply.
 */
void Object::animation( void (*anim_func)( TransCache &arg ) ) {
  anim_func( this->_trans );
  Object::propagate();
}

/**
 * Scene-graph changes are not automatically applied to children.
 * For efficiency reasons, you need to call propagate() manually.
 */
void Object::propagate( void ) {

  //fprintf( stderr, "\n" );
  //fprintf( stderr, "propagate called on %s\n", _name.c_str() );

  std::list< Object* >::iterator it;

  //std::cerr << "Calling CALCCTM:\n";
  //Update my Object's CTM...
  this->_trans.CalcCTM();

  //send my OTM as the PTM to all of my children.
  for ( it = _list.begin(); it != _list.end(); ++it ) {
    (*it)->_trans.PTM( this->_trans.OTM() );
    //Tell that child to update his CTM and propagate.
    (*it)->propagate();
  }

  //std::cerr << "{" << _name << "::OTM:" << this->_trans.OTM() << "}\n";

}

/**
 * Obtain the vec4 representative of the Object's current position in space.
 * @return vec4 representing the Object's position in space.
 */
vec4 Object::position( void ) const {

  mat4 theOTM = this->_trans.OTM();

  return vec4( theOTM[0][3], theOTM[1][3], theOTM[2][3], 1.0 );
}

/**
 * Retrieve a pointer to this object's morph target.
 * @return An Object pointer to the morph target.
 */
Object* Object::morphTarget( void ) const {

  return _morphTarget;

}

/**
 * Instantiate a new morphing target.
 * @param shader The shader to use for the new morphing target.
 * NOT USED for rendering the object, but Objects cannot be instantiated
 * without a shader, so here it is.
 *
 * @return A pointer to the newly created target.
 */
Object* Object::genMorphTarget( GLuint shader ) {

  Object *obj = new Object( this->_name + "_morph", shader );
  _morphTarget = obj;
  return obj;

}

/**
 * Retrieve the morph Percentage of this object.
 * @return The morph percentage, as a float.
 */
float Object::morphPercentage( void ) const {

  return _morphPercentage;
}

/**
 * Set the morph percentage of this Object.
 * @param newPercentage The new morphing percentage.
 */
void Object::morphPercentage( const float newPercentage ) {
  _morphPercentage = newPercentage;
}

/**
 * Obliterate the morph target for this object.
 */
void Object::destroyMorphTarget( void ) {

  if ( _morphTarget != NULL ) {
    delete _morphTarget;
    _morphTarget = NULL;
  }

}

/**
 * Retrieve the number of _vertices this object has.
 * @return An integer representing the number of vertices the object has.
 */
int Object::numberOfPoints( void ) {
  return _vertices.size();
}

